Client/Frontend/Browser/Punycode.swift

```diff
@@ -0,0 +1,209 @@
+                    output.append(toValue(code))
+                    q = (q - t) / (base - t)
+                    k += base
+                }
+                output.append(toValue(q))
+                bias = self.adapt(delta, numPoints: h + 1, firstTime: h == b)
+                delta = 0
+                h += 1
+            }
+            delta += 1
+            n += 1
+        }
+        return output
+    }
+
+    private func decode(punycode: String) -> String {
+        let base = 36
+        let tMin = 1
+        let tMax = 26
+        let initialBias = 72
+        let initialN: Int = 128 // 0x80
+        let delimiter: Character = "-"; // '\x2D'
+        var input = [Character](punycode.characters)
+        var output = [Character]()
+        var i = 0
+        var n = initialN
+        var bias = initialBias
+        var pos = 0
+        if let ipos = input.indexOf(delimiter) {
+            pos = ipos
+            output.appendContentsOf(input[0 ..< pos++])
+        }
+        var outputLength = output.count
+        let inputLength = input.count
+        while pos < inputLength {
+            let oldi = i
+            var w = 1
+            var k = base
+            while true {
+                let digit = toIndex(input[pos++])
+                i += digit * w
+                let t = max(min(k - bias, tMax), tMin)
+                if digit < t {
+                    break
+                }
+                w = w * (base - t)
+                k += base
+            }
+            bias = adapt(i - oldi, numPoints: ++outputLength, firstTime: (oldi == 0))
+            n = n + i / outputLength
+            i = i % outputLength
+            output.insert(Character(UnicodeScalar(n)), atIndex: i)
+            i += 1
+        }
+        return String(output)
+    }
+
+    private func isValidUnicodeScala(s: String) -> Bool {
+        let initialN: Int = 128 // 0x80
+        for c in s.unicodeScalars{
+            let ci = Int(c.value)
+            if ci >= initialN {
+                return false
+            }
+        }
+        return true
+    }
+
+    private func isValidPunycodeScala(s: String) -> Bool {
+        return s.hasPrefix("xn--")
+    }
+
+    public func utf8HostToAscii() -> String {
+        let s = self
+        let prefixPunycode = "xn--"
+        if isValidUnicodeScala(s) {
+            return s
+        }
+        var labels = s.componentsSeparatedByString(".")
+        var i : Int = 0
+        for part in labels{
+            if isValidUnicodeScala(part) != true{
+                let a = encode(part)
+                labels[i] = prefixPunycode + a
+            }
+            i += 1
+        }
+        let resultString = labels.joinWithSeparator(".")
+        return resultString
+    }
+
+    public func asciiHostToUTF8() -> String {
+        let s = self
+        var labels = s.componentsSeparatedByString(".")
+        var index : Int = 0
+        for part in labels {
```

Comment:
Normal way to do this is `for (i, part) in labels.enumerate { ... }`. That way, you don't need to maintain the index separately.